src/blog/fearless-structured-concurrency.vmd

```diff
@@ -0,0 +1,542 @@
+---
+title: Combining Fearless Concurrency with Structured Concurrency
+author: Evan Ovadia
+date: Jan 8, 2022
+realm: blog
+path: blog/fearless-structured-concurrency
+layout: annotated
+namespace: c-blog m-annotated
+---
+
+With fearless structured concurrency, we can add multi-threading to existing code with a single keyword, with no risk of data races!
+
+
+Summary:
+
+ * "Structured" concurrency makes multi-threading much *easier*.
+    * Threads can access data from the containing scope!
+ * "Fearless" concurrency *prevents data races* between threads.
+ * *Structured fearless concurrency combines the best of both worlds.*
+ * Vale's `parallel` keyword makes it happen!
+
+
+# We Want Structured Concurrency!
+
+
+Structured concurrency is a lot of fun.
+
+
+With structured concurrency, you can launch multiple threads at once, to run your calculations in parallel, for a big speedup. And often, it can be done with a tiny adjustment.
+
+
+Imagine we had this simple program that calculated x^3 for some values of x:
+
+```c
+#include <math.h>
+#include <stdio.h>
+
+void main() {
+  int exponent = 3;
+  int results[5];
+
+  // Calculate some exponents
+  for (int i = 0; i < 5; i++) {
+   results[i] = pow(i, exponent);
+  }
+
+  // Print out the results
+  for (int i = 0; i < 5; i++) {
+   printf("%d to the %d'th power is %d!\n", i, exponent, results[i]);
+  }
+}
+```stdout
+0^3 is 0!
+1^3 is 1!
+2^3 is 8!
+3^3 is 27!
+4^3 is 64!
+```
+
+
+
+If `pow` was much more expensive, we might want to run those calculations in parallel.
+
+It's very tedious to run it in C. We'd have to wrap our `results[i] = pow(i, exponent);` line in an entire new function, and use `pthread_create`, like so:
+
+
+```c
+#include <math.h>
+#include <stdio.h>
+#include <pthread.h>
+
+typedef struct {
+  int exponent;
+  int i;
+  int* results;
+} MyThreadArgs;
+
+void *my_thread_main(void* args_raw) {
+  MyThreadArgs* args = (MyThreadArgs*)args_raw;
+  int exponent = args->exponent;
+  int i = args->i;
+  int* results = args->results;
+
+  results[i] = pow(i, exponent); // some expensive calculation
```

Comment:
Use consistent capitalization of comments

Reply:
Done